src/inference/include/openvino/runtime/async_infer_queue.hpp

```diff
@@ -0,0 +1,124 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+/**
+ * @brief A header file that provides wrapper classes for infer requests and callbacks.
+ *
+ * @file openvino/runtime/async_infer_queue.hpp
+ */
+#pragma once
+
+#include <condition_variable>
+#include <map>
+#include <mutex>
+#include <queue>
+#include <string>
+#include <vector>
+
+#include "openvino/runtime/common.hpp"
+#include "openvino/runtime/executable_network.hpp"
+#include "openvino/runtime/infer_request.hpp"
+#include "openvino/runtime/tensor.hpp"
+
+namespace ov {
+namespace runtime {
+
+/**
+ * @brief This is an interface of asynchronous infer request
+ *
+ * It can throw exceptions safely for the application, where it is properly handled.
+ *
+ * @note AsyncInferQueue provides simple and unifed interface for multiple asynchronous
+ * calls to OpenVINO's InferRequests.
+ */
+class OPENVINO_RUNTIME_API AsyncInferQueue {
+public:
+    AsyncInferQueue() = delete;
+
+    /**
+     * @brief Basic constructor. Creates InferRequests based on given ExecutableNetwork.
+     *
+     * @param net ExecutableNetwork object which will be base for creating of InferRequests.
+     * @param jobs Number of InferRequests to be created. If equals to @p 0 AsyncInferQueue
+     * will automatically try to choose number of jobs.
+     */
+    AsyncInferQueue(ExecutableNetwork& net, size_t jobs = 0);
+
+    // ~AsyncInferQueue() {
+    //     _requests.clear();
+    // }
+
+    /**
+     * @brief Starts asynchronous inference on next avaiable job. Blocking call if there are no free jobs.
+     *
+     * @param userdata Data passed to InferRequest's callback. Can be anything that matches ov::Any.
+     */
+    void start_async(const ov::Any userdata = nullptr);
```

Comment:
Let add `bool try_start_async()` non blocking method. It will return true if async infer is started and false if there is no workers in pool

Reply:
I do not think that this is a good approach in first place. `AsyncInferQueue` has been designed to contain best of two worlds, highly-efficient C++ and user-friendly Python. The design has been accepted and it is not necessary to introduce new features like this one.

Reply:
But it would be placed into a common C++ API and it won't be used only from Python API.
try_start_async() - has trivial implementation but it can allow more advanced users to use this queue for more complex cases. 
For example, it can be used for complex flow graph with many input data producers and inference result consumers.